sourcegraph/sourcegraph
# faster webpack builds & rebuilds, upgrade old JS libs

Motivation: Make it faster to iterate on the web app.

* Upgrades Webpack and related libs
* Tweaks some settings as described in the commits (e.g., faster sourcemap generation on dev)
* Enables hot module reloading, so new JS and CSS is injected in-place
* Eliminates the `.css` files and loads styles via JS (speeds up builds, simplifies Webpack config, and in the future can let us do code splitting better)
* Adds a progress bar to Webpack

Introduces a requirement of Node v5+ so that we can use ES6 `const` in our webpack.config.js. This is not strictly necessary, but we all should have linter warnings in our editors for `var`, and it's annoying if the Webpack config file yields linter errors.

Up **SOON** on staging at https://staging4.sourcegraph.com

#### Review plan

A LOT of code changed, mostly npm deps. You're best off looking commit-by-commit. Ignore the commit diffs for the "upgrade" commits; you can just look at the npm commands I ran.

#### Test plan

1. Check out this branch
2. Quit Webpack and restart it. It should show a progress bar and take a LOT less time (~25s not ~70s).
3. Open http://localhost:3080 in your browser.
4. Make some edits to JS and CSS:
  * Add the string "aaaaaaa" to the `GlobalNav` `render()` method to make sure it shows up without you needing to reload the page
  * Change the GlobalNav border by adding `border: solid 2px red !important;` to some GlobalNav.css classes. Make sure it shows up.
  * Insert a JavaScript syntax error and make sure it is shown in the JS console log.
  * Insert a CSS syntax error and make sure it is shown in the browser console log.
5. Ensure you can run `npm test` from the `app/` dir successfully.

#### Acceptance criteria

Need a few people to make sure it works for them on various environments. I listed some people, but if you are not listed and want to help push this along, go ahead and change the name and check the box if you can verify it works:

* [x] Linux (Ubuntu 15.04, Node v5.11) @sqs 
* [ ] Linux  @beyang
* [ ] Linux VM @alexsaveliev 
* [x] macOS @rothfels @KattMingMing @neelance 


#### Open issues

* [ ] The CI build on branch staging4 is flaky. The flow executable is sometimes killed. It wasn't working at all using the `./node_modules/.bin/flow` path in the app/package.json `scripts.check` field, but it passed once when I made the path just `flow`.
* [x] **FIXED by installing redbox-react (in-UI error reporter)** ~~Sometimes, it just reports an error like the following, only mentioning `AppContainer` (which is part of the hot module reloading scaffolding code) and an unrelated ScrollBehaviorContainer `stop` error:~~
   ```
ScrollBehaviorContainer.js:64 Uncaught TypeError: Cannot read property 'stop' of undefinedcomponentWillUnmount @ ScrollBehaviorContainer.js:64ReactErrorUtils
invariant.js:38 Uncaught Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object. Check the render method of `AppContainer`.
   ```
